File: go/internal/service/node_service.go
package service

import (
	"context"
	"fmt"

	"github.com/hemanthpathath/flex-db/go/internal/repository"
)

// NodeService handles node business logic
type NodeService struct {
	repo         repository.NodeRepository
	nodeTypeRepo repository.NodeTypeRepository
}

// NewNodeService creates a new NodeService
func NewNodeService(repo repository.NodeRepository, nodeTypeRepo repository.NodeTypeRepository) *NodeService {
	return &NodeService{repo: repo, nodeTypeRepo: nodeTypeRepo}
}

// Create creates a new node
func (s *NodeService) Create(ctx context.Context, tenantID, nodeTypeID, data string) (*repository.Node, error) {
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	if nodeTypeID == "" {
		return nil, fmt.Errorf("node_type_id is required")
	}

	// Validate that the node type belongs to the same tenant
	nodeType, err := s.nodeTypeRepo.GetByID(ctx, tenantID, nodeTypeID)
	if err != nil {
		return nil, fmt.Errorf("invalid node_type_id: node type not found or does not belong to this tenant")
	}
	if nodeType.TenantID != tenantID {
		return nil, fmt.Errorf("invalid node_type_id: node type does not belong to this tenant")
	}

	node := &repository.Node{
		TenantID:   tenantID,
		NodeTypeID: nodeTypeID,
		Data:       data,
	}

	return s.repo.Create(ctx, node)
}

// GetByID retrieves a node by ID
func (s *NodeService) GetByID(ctx context.Context, tenantID, id string) (*repository.Node, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	return s.repo.GetByID(ctx, tenantID, id)
}

// Update updates an existing node
func (s *NodeService) Update(ctx context.Context, tenantID, id, data string) (*repository.Node, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}

	node, err := s.repo.GetByID(ctx, tenantID, id)
	if err != nil {
		return nil, err
	}

	if data != "" {
		node.Data = data
	}

	return s.repo.Update(ctx, node)
}

// Delete deletes a node
func (s *NodeService) Delete(ctx context.Context, tenantID, id string) error {
	if id == "" {
		return fmt.Errorf("id is required")
	}
	if tenantID == "" {
		return fmt.Errorf("tenant_id is required")
	}
	return s.repo.Delete(ctx, tenantID, id)
}

// List retrieves nodes with pagination and optional filtering
func (s *NodeService) List(ctx context.Context, tenantID, nodeTypeID string, pageSize int32, pageToken string) ([]*repository.Node, *repository.ListResult, error) {
	if tenantID == "" {
		return nil, nil, fmt.Errorf("tenant_id is required")
	}

	opts := repository.ListOptions{
		PageSize:  int(pageSize),
		PageToken: pageToken,
	}
	return s.repo.List(ctx, tenantID, nodeTypeID, opts)
}

File: python/app/service/node_service.py
"""
Node service implementation.
"""

from typing import List, Optional, Tuple

from app.repository import Node, NodeRepository, NodeTypeRepository, ListOptions, ListResult


class NodeService:
    """Node business logic service."""

    def __init__(self, repo: NodeRepository, node_type_repo: NodeTypeRepository):
        self.repo = repo
        self.node_type_repo = node_type_repo

    async def create(self, node_type_id: str, data: str) -> Node:
        """Create a new node."""
        if not node_type_id:
            raise ValueError("node_type_id is required")

        # Validate that the node type exists (repository is already scoped to tenant database)
        node_type = await self.node_type_repo.get_by_id(node_type_id)

        node = Node(
            tenant_id="",  # Not stored in tenant database
            node_type_id=node_type_id,
            data=data,
        )
        return await self.repo.create(node)

    async def get_by_id(self, id: str) -> Node:
        """Retrieve a node by ID."""
        if not id:
            raise ValueError("id is required")
        return await self.repo.get_by_id(id)

    async def update(self, id: str, data: str) -> Node:
        """Update an existing node."""
        if not id:
            raise ValueError("id is required")

        node = await self.repo.get_by_id(id)

        if data:
            node.data = data

        return await self.repo.update(node)

    async def delete(self, id: str) -> None:
        """Delete a node."""
        if not id:
            raise ValueError("id is required")
        await self.repo.delete(id)

    async def list(
        self,
        node_type_id: Optional[str],
        page_size: int,
        page_token: str
    ) -> Tuple[List[Node], ListResult]:
        """Retrieve nodes with pagination and optional filtering."""
        opts = ListOptions(page_size=page_size, page_token=page_token)
        return await self.repo.list(node_type_id, opts)
